sklearn/preprocessing/data.py

```diff
@@ -54,41 +58,82 @@
 )
 
 
-def _mean_and_std(X, axis=0, with_mean=True, with_std=True):
-    """Compute mean and std deviation for centering, scaling.
+def _incremental_mean_and_var(new_X, axis=0, last_mean=.0, last_var=None,
+                              last_n_samples_seen=0):
+    """Compute incremental mean and var for centering, scaling.
 
-    Zero valued std components are reset to 1.0 to avoid NaNs when scaling.
+    last_mean and last_var are the statistics computed at the last step by this
+    function. Both must be initilized to 0.0. In case no scaling is required
+    (caller has with_std=False), last_var can be None. The mean is always
+    required and returned (even if caller has with_mean=True) because it is
+    necessary for the calculation of the var. last_n_samples_seen is the number
+    of samples encountered until now.
     """
-    X = np.asarray(X)
-    Xr = np.rollaxis(X, axis)
+    new_Xr = np.rollaxis(new_X, axis)
 
-    if with_mean:
-        mean_ = Xr.mean(axis=0)
+    if last_var is not None:
+        updated_mean, updated_var, updated_n = \
+            _batch_mean_variance_update(new_Xr, last_mean=last_mean,
```

Comment:
It seems to me that there is code duplication while finding the mean?

Do you think it would be better to have a with_var keyword in _batch_mean_variance_update and do away with this function altogether?